dom.bs

```diff
@@ -1924,46 +1947,114 @@ is [=AbortSignal/aborted=]; otherwise false.
 them. For instance, if the operation has already completed.
 
-<p>To <dfn export for=AbortSignal>signal abort</dfn>, given an {{AbortSignal}} object
-<var>signal</var> and an optional <var>reason</var>, run these steps:
+<hr>
+
+<p>An {{AbortSignal}} object is <dfn export for="AbortSignal">aborted</dfn> when its
+[=AbortSignal/abort reason=] is not undefined.
+
+<div algorithm>
+<p>To <dfn export for=AbortSignal>add</dfn> an algorithm <var>algorithm</var> to an {{AbortSignal}}
+object <var>signal</var>:
+
+<ol>
+ <li><p>If <var>signal</var> is [=AbortSignal/aborted=], then return.
+
+ <li><p><a for=set>Append</a> <var>algorithm</var> to <var>signal</var>'s
+ <a for=AbortSignal>abort algorithms</a>.
+</ol>
+</div>
+
+<div algorithm>
+<p>To <dfn export for=AbortSignal>remove</dfn> an algorithm <var>algorithm</var> from an
+{{AbortSignal}} <var>signal</var>, <a for=set>remove</a> <var>algorithm</var> from
+<var>signal</var>'s <a for=AbortSignal>abort algorithms</a>.
+</div>
+
+<div algorithm>
+<p>To <dfn for=AbortSignal>signal abort</dfn>, given an {{AbortSignal}} object <var>signal</var> and
+an optional <var>reason</var>:
 
 <ol>
  <li><p>If <var>signal</var> is [=AbortSignal/aborted=], then return.
 
  <li><p>Set <var>signal</var>'s [=AbortSignal/abort reason=] to <var>reason</var> if it is given;
  otherwise to a new "{{AbortError!!exception}}" {{DOMException}}.
 
- <li><p><a for=set>For each</a> <var>algorithm</var> in <var>signal</var>'s
+ <li><p><a for=set>For each</a> <var>algorithm</var> of <var>signal</var>'s
  [=AbortSignal/abort algorithms=]: run <var>algorithm</var>.
 
  <li><p><a for=set>Empty</a> <var>signal</var>'s <a for=AbortSignal>abort algorithms</a>.
 
  <li><p>[=Fire an event=] named {{AbortSignal/abort}} at <var>signal</var>.
+
+ <li><p><a for=set>For each</a> <var>dependentSignal</var> of <var>signal</var>'s
+ [=AbortSignal/dependent signals=], [=AbortSignal/signal abort=] on <var>dependentSignal</var> with
+ <var>signal</var>'s [=AbortSignal/abort reason=].
 </ol>
+</div>
 
-<p>A <var>followingSignal</var> (an {{AbortSignal}}) is made to
-<dfn export for=AbortSignal>follow</dfn> a <var>parentSignal</var> (an {{AbortSignal}}) by running
-these steps:
+<div algorithm>
+<p>To <dfn export>create a composite abort signal</dfn> from a list of {{AbortSignal}} objects
+<var>signals</var>, using <var>signalInterface</var>, which must be either {{AbortSignal}} or an
+interface that inherits from it, and a <var>realm</var>:
 
 <ol>
- <li><p>If <var>followingSignal</var> is [=AbortSignal/aborted=], then return.
+ <li><p>Let <var>resultSignal</var> be a <a for=/>new</a> object implementing
+ <var>signalInterface</var> using <var>realm</var>.
 
- <li><p>If <var>parentSignal</var> is [=AbortSignal/aborted=], then
- <a for=AbortSignal>signal abort</a> on <var>followingSignal</var> with <var>parentSignal</var>'s
- [=AbortSignal/abort reason=].
+ <li><p>Set <var>resultSignal</var>'s [=AbortSignal/composite=] to true.
```

Comment:
Should we only do this after iteration and only if _signals_ is not the empty list?

Not sure why we would return a composite signal if any of the dependent signals is already aborted for instance or if there are no dependent signals.

Reply:
I think setting it after checking aborted signals is probably okay, or at least functionally equivalent?

If it's an empty list, I think it should still be composite. If not, the signal could be added as a source for another composite signal, e.g.

```js
let signal1 = AbortSignal.any([]);
// If `signal1` is not composite, it would be added as a source for `signal2`, but
// it will never be aborted.
let signal2 = AbortSignal.any([signal1, realSignal]);
```

That would affect GC eligibility: `signal1` should be GC-able immediately and `signal2` would be GC-able when `realSignal` is GCed.

I think this maybe comes down to how SW/internal signals are handled. In https://github.com/whatwg/fetch/pull/1646#issuecomment-1532212040, I'm advocating for SW not creating a composite signal (or at least creating one with a source signal it controls) to avoid composite signals being aborted directly, which I think would be surprising (a hidden source!).

Reply:
I see, that makes sense. In that case setting this state upfront is clearer.

And `Request` objects creating empty composite signals makes sense as well as there is not always something controlling them and this would make that more explicit. Though maybe that means composite isn't completely the right term? Perhaps they are "uncontrolled" / "unowned" or some such?

Reply:
Moved setting the flag until after checking the input signals for being aborted.

Not sure this is the right framing, but I've been thinking of such signals as being 'composed of nothing' or 'dependent on the empty set (of signals)', which is inert but still composite/dependent (note: signals can reach this state if source signals are GCed). Would 'dependent' is better, or does that have the same problem? 'dependent' could be nice because it matches the 'dependent signals' variable.

I have a slight preference for dependent/composite because it mentions a relationship to the input signals, but I don't feel strongly. unowned/uncontrolled could work too, but I think an "create an uncontrolled abort signal given [empty set]" would be less clear/readable than "create an dependent abort signal given [empty set]".

Reply:
Dependent seems reasonable. Thanks!

Reply:
Great, done.